### ISSUE TITLE
Practices for performance control

### ISSUE THREAD

**Issue by jeluard** [status-im/ideas#22]
Idea: DEV#0022
Title: Practices for performance control
Status: Draft
Created: 2017-11-17

## Summary

A set of practices (tools, documents, methods) to help diagnose and prevent performance regressions

## Vision

User experience is paramount for user retention, especially in mobile environment.
Performance issues are hard to pinpoint without special tooling and trivial changes might introduce subtil performance regression.
We want to ensure we own performance profile of our products.

## Swarm Participants

* Lead contributor: @jeluard 
* Contributor: @themue 
* Contributor: @antdanchenko  
* Evaluator:

### Pinpoint areas of slow performance

Provide a list of tools with associated documentation to quickly identify sources of performance issues.
Specifically those performance issues might lie:

* at the react level (e.g. suboptimal re-frame usage, algorithmic complexity)
* at the go level (e.g. whisper overhead)

__NOTE__ Fixing those issues is not part of the scope

### Automatize prevention of performance regression

Simple change might introduce performance regressions. Automatizing the discovery of those regressions reduces the probability that they will be shipped.

### Improve perceived reactivity

Even when everything is optimized there can still be some inherent slowness, especially on older mobile.
Some UX technics can help improving the perceived performance, like:

* screen open fast
* `jank-free` scrolling
* smooth transition
* consistent loading and error handling
* animations?

### MVP

When it comes to performance optimizations, gut feeling is generally wrong.
Define procedures (tools and documentations) allowing to discover the root cause of performance issues.

Goal Date: _To be defined_
Description: a set of rules allowing to pinpoint performance issues

### Iteration #1 

Automatize performance regression. Some specific e2e tests will track performance of common scenario and will be integrated in Jenkins.
If customizable thresholds are crossed build fails.

Goal Date: _To be defined_
Description: improvements to our CI preventing performance regressions

### Iteration #2

Define precise guidelines and technics that must be implemented consistently to provide a smooth UX.

Goal Date: _To be defined_
Description: a set of practices to improve perceived reactivity

### Copyright
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

**Comment by themue** [comment]
Regarding _status-go_ see https://github.com/status-im/status-go/projects/5.

**Comment by oskarth** [comment]
@jeluard Great initiative! I'm interested. Don't want to pledge time until we've made more progress with offline inboxing though. Any idea roughly when you plan on starting on this?

**Comment by jeluard** [comment]
@oskarth Probably not before end of next week but let's start discussing!

**Comment by oskarth** [comment]
Relevant to this issue is this except from Rob Pike's 5 Rules of Programming:

```
Rule 1. You can't tell where a program is going to spend its time. Bottlenecks occur in surprising places, so don't try to second guess and put in a speed hack until you've proven that's where the bottleneck is.
Rule 2. Measure. Don't tune for speed until you've measured, and even then don't unless one part of the code overwhelms the rest.
```
(http://users.ece.utexas.edu/~adnan/pike.html)

**Comment by antdanchenko** [comment]
Please, see comment
It seems that the solution might be applied for both ideas

**Comment by jeluard** [comment]
This `idea` is currently on hold as it overlaps significantly the ongoing [idea #55](https://github.com/status-im/ideas/issues/55).
Its scope will be reconsidered once `idea #55` is done.

**Comment by jeluard** [comment]
Closing as it has been superseded by #55
